### PR TITLE
TYP: update for numpy-1.21.0

### PR DESCRIPTION
**Pull request by simonjayhawkins**

### CODE REVIEW

File: pandas/core/internals/construction.py [hunk 1]
```diff
@@ -142,7 +137,7 @@ def arrays_to_mgr(
 
 
 def rec_array_to_mgr(
-    data: MaskedRecords | np.recarray | np.ndarray,
+    data: np.ma.MaskedArray | np.recarray | np.ndarray,
```
Comment (jbrockmendel) [inline pandas/core/internals/construction.py:140]: why this change?

File: pandas/core/nanops.py [hunk 2]
```diff
@@ -1784,11 +1790,9 @@ def na_accum_func(values: ArrayLike, accum_func, *, skipna: bool) -> ArrayLike:
             # DatetimeArray/TimedeltaArray
             # TODO: have this case go through a DTA method?
             # For DatetimeTZDtype, view result as M8[ns]
+            values = cast("DatetimeArray | TimedeltaArray", values)
```
Comment (jbrockmendel) [inline pandas/core/nanops.py:1793]: i think just DatetimeArray (comment on L1790 should reflect this too); OK for follow-up.

**Comment by jbrockmendel** [comment]
The MaskedRecords change is the only one that looks questionable, everything else I think we can merge and cleanup in follow-ups.

File: pandas/core/internals/construction.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/core/internals/construction.py:140]: There's a 
```
    # fill if needed
    if isinstance(data, np.ma.MaskedArray):
        new_arrays = fill_masked_arrays(data, arr_columns)
```
in the body that implies that MaskedArray (which is a superclass?) is allowed?

File: pandas/core/nanops.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/core/nanops.py:1793]: actually it's probably best to remove the cast and keep the "fix later" ignore. A 3rd party EA subclass could satisfy `values.dtype.kind in ["m", "M"]` and would not have a `_simple_new` method as that's not a requirement of the EA api.

**Comment by simonjayhawkins** [comment]
> The MaskedRecords change is the only one that looks questionable

sure. will look at that in detail tomorrow. feel free to push changes here in the meantime.

**Comment by jreback** [comment]
may need to do a 1.2.6 with this unfortunately

**Comment by jreback** [comment]
@simonjayhawkins we need for 1.3 i assume.

File: pandas/core/internals/construction.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/internals/construction.py:140]: In all tests that get here, we have `isinstance(data, np.ma.mrecords.MaskedRecords)`.  AFAICT the check is for MaskedArray bc MaskedRecords requires a separate import that we are avoiding at import-time

**Comment by jbrockmendel** [comment]
just pushed restoring the MaskedRecords annotation

File: pandas/core/internals/construction.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/core/internals/construction.py]: `MaskedRecords`, `MaskedArray` and `recarray` are all subclass of `np.ndarray` so the annotation here is redundant and should just be `np.ndarray`

Since `MaskedArray` is a subclass of `np.ndarray`, it is a valid type to pass to `rec_array_to_mgr` and mypy is correct that we pass off `MaskedArray` classes and subclasses to `fill_masked_arrays`

> In all tests that get here

Isn't the point of mypy to catch the cases where we don't have tests?

but ok to leave for now. as we now have a cast and comment to highlight the issue

**Comment by simonjayhawkins** [comment]
> @simonjayhawkins we need for 1.3 i assume.

not strictly, but best not to diverge the ci before we release

> may need to do a 1.2.6 with this unfortunately

There are no behavioral code changes that indicate that there are issues with 1.2.5 and numpy 1.21

**Comment by simonjayhawkins** [comment]
`test_read_csv` failing on Windows py38_np18 is not down to changes here

> > may need to do a 1.2.6 with this unfortunately
> 
> There are no behavioral code changes that indicate that there are issues with 1.2.5 and numpy 1.21

we had testing on 1.2.x for numpy-dev, but if we do get reports of issues with 1.2.5 and numpy 1.21 we may well want to do a 1.2.6 to fix (and then backport these changes to 1.2.x also, 1.2.x is on mypy 0.782 but probably not an issue)

File: pandas/core/dtypes/cast.py [hunk 3]
```diff
@@ -107,6 +108,8 @@
 _int32_max = np.iinfo(np.int32).max
 _int64_max = np.iinfo(np.int64).max
 
+NumpyArrayT = TypeVar("NumpyArrayT", bound=np.ndarray)
```
Comment (jbrockmendel) [inline pandas/core/dtypes/cast.py:111]: why can't we just use np.ndarray?
